src/mod_hacker_news.erl

```diff
@@ -0,0 +1,154 @@
+config_spec() ->
+    #section{
+       items = #{<<"topstories_url">> => #option{type = string, validate = non_empty},
+                 <<"topstories_total">> => #option{type = integer, validate = positive},
+                 <<"topstories_interval">> => #option{type = integer, validate = positive},
+                 <<"topstories_retry">> => #option{type = integer, validate = positive}
+                }
+      }.
+
+-spec hook_handlers() -> list().
+hook_handlers() ->
+    [{user_send_packet, user_send, 100}].
+
+%%
+%% user_send_packet handler
+%%
+
+user_send(Acc, _JID, _From, _Packet) ->
+    ejabberd_c2s:run_remote_hook(self(), ?MODULE, init),
+    Acc.
+
+%%
+%% gen_iq_handler handlers
+%%
+
+-spec process_iq(jid:jid(), mongoose_acc:t(), jid:jid(), jlib:iq()) -> {mongoose_acc:t(), jlib:iq()} | error().
+process_iq(From, _To, Acc, #iq{xmlns = ?NS_HACKER_NEWS, type = get} = IQ) ->
+    IQResp = case lists:member(From#jid.lserver, ?MYHOSTS) of
+        true ->
+            process_local_iq(IQ);
+        false ->
+            iq_error(IQ, [mongoose_xmpp_errors:item_not_found()])
+    end,
+    {Acc, IQResp};
+process_iq(_From, _To, Acc, #iq{} = IQ) ->
+    {Acc, iq_error(IQ, [mongoose_xmpp_errors:bad_request()])}.
+
+process_local_iq(IQ) ->
+    try
+        create_topstories_response(IQ)
+    catch
+        _:_ ->
+            iq_error(IQ, [mongoose_xmpp_errors:internal_server_error()])
+    end.
+
+iq_error(IQ, SubElements) when is_list(SubElements) ->
+    IQ#iq{type = error, sub_el = SubElements}.
+
+create_topstories_response(IQ) ->
+    Topstories = get_topstories(),
+    IQ#iq{type = result,
+          sub_el = [#xmlel{name = <<"items">>,
+                           attrs = [{<<"xmlns">>, ?NS_HACKER_NEWS}],
+                           children = [topstories_to_xmlel(jiffy:encode(Topstories))]}]}.
+
+topstories_to_xmlel(Data) ->
+    #xmlel{name = <<"topstories">>,
+           children = [#xmlcdata{content = Data}]}.
+
+%%
+%% API
+%%
+
+-spec get_topstories() -> Res :: list().
+get_topstories() ->
+    persistent_term:get(?CACHE_TOPSTORIES, []).
+
+-spec cache_topstories(Data :: maps:map()) -> ok.
+cache_topstories(#{url := Url, total := Total,
+                   interval := Interval,
+                   retry := Retry, domain := Domain} = Opts) ->
+    _ = mongoose_metrics:update(Domain, [?MODULE, ?METRIC_TOPSTORIES], 1),
+    _ = timer:apply_after(Interval, ?MODULE, ?FUNCTION_NAME, [Opts]),
+    case maybe_retry(Url, Retry) of
+        error ->
+            ok;
+        {ok, Res} ->
+            ok = persistent_term:put(?CACHE_TOPSTORIES, lists:sublist(Res, Total))
+    end.
+
+-spec maybe_retry(Url :: list(), Retry :: integer()) -> {ok, Res :: list()} | error.
+maybe_retry(_, 0) ->
+    error;
+maybe_retry(Url, Retry) ->
+    try 
+        {ok, {{_, 200, _}, _, Body}} = httpc:request(Url),
+        Res = jiffy:decode(Body),
+        {ok, Res}
+    catch
+        _:_ ->
```

Comment:
It might be good to log the nature of the error, especially when we run out of retries.

Reply:
True, totally agree.